TAO/orbsvcs/orbsvcs/Event/ECG_Mcast_EH.cpp

```diff
@@ -97,30 +98,35 @@ TAO_ECG_Mcast_EH::shutdown (void)
   // Indicates that we are in a shutdown state.
   this->receiver_ = 0;
 
-  // Deregister from reactor, close and clean up sockets.
-  size_t const subscriptions_size = this->subscriptions_.size ();
-  for (size_t i = 0; i != subscriptions_size; ++i)
-    {
-      (void) this->reactor ()->remove_handler (
-                               this->subscriptions_[i].dgram->get_handle (),
-                               ACE_Event_Handler::READ_MASK);
-      (void) this->subscriptions_[i].dgram->close();
-      delete this->subscriptions_[i].dgram;
-    }
-  this->subscriptions_.size (0);
+  {
+    ACE_GUARD_RETURN (TAO_SYNCH_MUTEX, ace_mon, this->lock_, -1);
+    // Deregister from reactor, close and clean up sockets.
+    size_t const subscriptions_size = this->subscriptions_.size ();
+    for (size_t i = 0; i != subscriptions_size; ++i)
+      {
+        (void) this->reactor ()->remove_handler (
+                                 this->subscriptions_[i].dgram->get_handle (),
+                                 ACE_Event_Handler::READ_MASK);
+        (void) this->subscriptions_[i].dgram->close();
+        delete this->subscriptions_[i].dgram;
+      }
+    this->subscriptions_.size (0);
+  }
 
   return 0;
 }
 
 int
 TAO_ECG_Mcast_EH::handle_input (ACE_HANDLE fd)
 {
+  ACE_GUARD_RETURN (TAO_SYNCH_MUTEX, ace_mon, this->lock_, -1);
   size_t const subscriptions_size = this->subscriptions_.size ();
   for (size_t i = 0; i != subscriptions_size; ++i)
     {
       ACE_SOCK_Dgram_Mcast *socket = this->subscriptions_[i].dgram;
       if (socket->get_handle () == fd)
         {
+          this->lock_.release ();
```

Comment:
You should use   ace_mon.release()  instead of lock_.release() since ace_mon is currently in control of the lock status.